pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -0,0 +1,310 @@
+
+	controller := &Handler{
+		ctx:                  ctx,
+		longhornSettings:     longhornSettings,
+		longhornSettingCache: longhornSettings.Cache(),
+		longhornVolumes:      longhornVolumes,
+		longhornVolumeCache:  longhornVolumes.Cache(),
+		prometheus:           prometheus,
+		prometheusCache:      prometheus.Cache(),
+		settings:             settings,
+		settingsController:   settings,
+		deployments:          deployments,
+		deploymentCache:      deployments.Cache(),
+		managedCharts:        managedCharts,
+		managedChartCache:    managedCharts.Cache(),
+	}
+
+	settings.OnChange(ctx, ControllerName, controller.OnStorageNetworkChange)
+	return nil
+}
+
+// webhook needs check if VMs are off
+// webhook needs check if nad is existing
+func (h *Handler) OnStorageNetworkChange(key string, setting *harvesterv1.Setting) (*harvesterv1.Setting, error) {
+	if setting == nil || setting.DeletionTimestamp != nil || setting.Name != settings.StorageNetworkName {
+		return nil, nil
+	}
+
+	if setting.Value == h.getLonghornStorageNetwork() {
+		// check if we need to restart monitoring pods
+		if !h.checkPodStatusAndStart() {
+			h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+		}
+
+		// finish
+		return nil, nil
+	}
+
+	logrus.Infof("storage network change:%s", setting.Value)
+
+	// if replica eq 0, skip
+	// save replica to annotation
+	// set replica to 0
+	if !h.checkPodStatusAndStop() {
+		logrus.Infof("Requeue to check pod status again")
+		h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+		return nil, nil
+	}
+
+	logrus.Infof("all pods are stopped")
+
+	// check volume detach before put LH settings
+	if !h.checkLonghornVolumeDetach() {
+		logrus.Infof("still has attached volume")
+		h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+		return nil, nil
+	}
+
+	logrus.Infof("all volumes are detached")
+	logrus.Infof("update Longhorn settings")
+	// push LH setting
+	if err := h.updateLonghornStorageNetwork(setting.Value); err != nil {
+		logrus.Warnf("Update Longhorn setting error %v", err)
+	}
+	h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+
+	return nil, nil
+}
+
+// true: all detach
+func (h *Handler) checkLonghornVolumeDetach() bool {
+	if volumes, err := h.longhornVolumeCache.List(util.LonghornSystemNamespaceName, labels.Everything()); err == nil {
+		for _, volume := range volumes {
+			logrus.Infof("volume state: %v", volume.Status.State)
+			if volume.Status.State != "detached" {
+				return false
+			}
+		}
+		return true
+	} else {
+		logrus.Warnf("volume error %v", err)
+	}
+	return true
+}
+
+// check Pod status, if all pods are start, return true
+func (h *Handler) checkPodStatusAndStart() bool {
+	allStarted := true
+
+	// check prometheus cattle-monitoring-system/rancher-monitoring-prometheus replica
+	if prometheus, err := h.prometheusCache.Get(CattleMonitoringSystemNamespace, RancherMonitoringPrometheus); err == nil {
```

Comment:
When err != nil,  checkPodStatusAndStart will return true, no error message, no retry.

Reply:
i will check it later